pkg/efivarfs/varfs.go

```diff
@@ -0,0 +1,243 @@
+	"path/filepath"
+	"sort"
+	"syscall"
+
+	"github.com/u-root/u-root/pkg/guid"
+	"golang.org/x/sys/unix"
+)
+
+func init() {
+	if !probeEfivarfs() {
+		return
+	}
+}
+
+// EfiVarFs is the path to the efivarfs mount point
+const EfiVarFs = "/sys/firmware/efi/efivars/"
+
+var vars varsBackend = varfsBackend{}
+
+type varsBackend interface {
+	Get(name string, guid guid.GUID) (VariableAttributes, []byte, error)
+	Set(name string, guid guid.GUID, attrs VariableAttributes, data []byte) (bool, error)
+	Delete(name string, guid guid.GUID) error
+	List() ([]VariableDescriptor, error)
+}
+
+type varfsBackend struct{}
+
+// Get reads the contents of an efivar if it exists and has the necessary permission
+func (v varfsBackend) Get(name string, guid guid.GUID) (VariableAttributes, []byte, error) {
+	path := filepath.Join(EfiVarFs, fmt.Sprintf("%s-%s", name, guid))
+	f, err := openVarfsFile(path, os.O_RDONLY, 0)
+	switch {
+	case os.IsNotExist(err):
+		return 0, nil, ErrVarNotExist
+	case os.IsPermission(err):
+		return 0, nil, ErrVarPermission
+	case err != nil:
+		return 0, nil, err
+	}
+	defer f.Close()
+
+	var attrs VariableAttributes
+	err = binary.Read(f, binary.LittleEndian, &attrs)
+	if err != nil {
+		if err == io.EOF {
+			return 0, nil, ErrVarNotExist
+		}
+		return 0, nil, err
+	}
+
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return 0, nil, err
+	}
+	return attrs, data, nil
+}
+
+// Set modifies a given efivar with the provided contents
+func (v varfsBackend) Set(name string, guid guid.GUID, attrs VariableAttributes, data []byte) (bool, error) {
+	path := filepath.Join(EfiVarFs, fmt.Sprintf("%s-%s", name, guid))
+	flags := os.O_WRONLY | os.O_CREATE
+	if attrs&AttributeAppendWrite != 0 {
+		flags |= os.O_APPEND
+	}
+
+	read, err := openVarfsFile(path, os.O_RDONLY, 0)
+	switch {
+	case os.IsNotExist(err):
+	case os.IsPermission(err):
+		return false, ErrVarPermission
+	case err != nil:
+		return false, err
+	default:
+		defer read.Close()
+
+		restoreImmutable, err := makeVarFileMutable(read)
+		switch {
+		case os.IsPermission(err):
+			return false, ErrVarPermission
+		case err != nil:
+			return false, err
+		}
+		defer restoreImmutable()
+	}
+
+	write, err := openVarfsFile(path, flags, 0644)
+	switch {
+	case os.IsPermission(err):
+		pe, ok := err.(*os.PathError)
+		if !ok {
+			return false, err
+		}
+		if pe.Err == syscall.EACCES {
+			// open will fail with EACCES if we lack the privileges
+			// to write to the file or the parent directory in the
+			// case where we need to create a new file. Don't retry
+			// in this case.
+			return false, ErrVarPermission
+		}
+
+		// open will fail with EPERM if the file exists but we can't
+		// write to it because it is immutable. This might happen as a
+		// result of a race with another process that might have been
+		// writing to the variable or may have deleted and recreated
+		// it, making the underlying inode immutable again. Retry in
+		// this case.
+		return true, ErrVarPermission
```

Comment:
ah, ok, I see a true here :-)
consider going this
ErrRetry = errors.New("EFIVarfs retry error")
or some such and then callers can just test for that.

Reply:
Ah that's a good idea, will do